python/orchestrator/health_status_monitor.py

```diff
@@ -0,0 +1,162 @@
+                   for component in target_components)
+
+    def __all_have_same_state(self, target_components):
+        """"
+        helper function to check the current local states of all components.
+        Returns boolean value indicating whether all current local
+        states are same.
+        """
+        return all(
+            component.current_state == target_components[0].current_state
+            for component in target_components)
+
+    def validate_local_states(self):
+        '''
+        checks the current status and the current state of each component.
+        If all have the same current state and status as UP
+        then, it returns a tuple of proxies of all components
+        and the componenets which have states (e.g. Orchestrator,
+        Application Companions).
+        Otherwise, it returns a tuple of None.
+        '''
+        # fetch all components from registry
+        all_components = self.__service_registry_manager.find_all()
+        components_with_states = []
+        if self.__all_status_up(all_components):
+            self.__logger.debug('all components are UP.')
+            # Filter the components which do not have states
+            # such as Command and Steering Service
+            for component in all_components:
+                if component.current_state is not None:
+                    components_with_states.append(component)
+            self.__logger.debug(f'all componenets: {all_components}; '
+                                f'with states: {components_with_states}')
+            if self.__all_have_same_state(components_with_states):
+                return all_components, components_with_states
+            else:
+                return None, None
+        else:
+            # find which components are DOWN
+            components_not_running = []
+            for component in all_components:
+                self.__logger.debug(f'{component.name} status: '
+                                    f'{component.current_status}.')
+                if component.current_status == SERVICE_COMPONENT_STATUS.DOWN:
+                    components_not_running.append(component)
+                    return None, None
+
+    @property
+    def keep_monitoring(self): return self.__keep_monitoring
+
+    def finalize_monitoring(self):
+        # stop monitoring
+        self.__keep_monitoring = False
+
+    def __monitor_health_status(self):
+        '''
+        Monitors the current status and the current state of each component.
+        Raises an alarm signal, if local states are in-consistent.
+        '''
+        counter = self.__counter
+        while self.keep_monitoring:
+            # local states are not the same, or statuses are not up.
+            if self.validate_local_states() == (None, None):
+                counter = counter - 1
+                if not counter:
+                    self.__is_confirmed = True
+                if self.__is_confirmed:
```

Comment:
I have a hard time understanding this while loop. 
This might be due to the name of the variable __is_confirmed.
Could you add some source code comment with regards to the intent of the code?